src/cipher/encrypt.go

```diff
@@ -0,0 +1,173 @@
+	dataHash := SumSHA256(data)
+	length := make([]byte, lenPrefixSize)
+	binary.PutUvarint(length, uint64(len(data)+32)) // the length including the data hash
+	var pbuf bytes.Buffer
+	pbuf.Write(length)
+	pbuf.Write(dataHash[:])
+	pbuf.Write(data)
+	pdata := pbuf.Bytes()
+
+	// split the data into 256 bit blocks(pad with null to 32 bytes)
+	l := len(pdata) // hash length + data length
+	n := l / blockSize
+	var blocks [][blockSize]byte
+	for i := 0; i < n; i++ {
+		var b [blockSize]byte
+		copy(b[:], pdata[i*blockSize:(i+1)*blockSize])
+		blocks = append(blocks, b)
+	}
+
+	// append last block if exist
+	if l%blockSize > 0 {
+		b := [blockSize]byte{}
+		copy(b[:], pdata[n*blockSize:])
+		blocks = append(blocks, b)
+	}
+
+	nonce := RandByte(blockSize)
+	var encryptedBlocks []string
+	// encode the blocks
+	for i := range blocks {
+		// hash(password, hash(index, hash(nonce)))
+		h := hashPwdIndexNonce(password, int64(i), nonce)
+		bh := SHA256(blocks[i])
+		encryptedHash := bh.Xor(h)
+		// encode the encrypted hash in base64
+		s := base64.StdEncoding.EncodeToString(encryptedHash[:])
+		encryptedBlocks = append(encryptedBlocks, s)
+	}
+
+	encryptedData := strings.Join(encryptedBlocks, ",")
+	nonceAndDataBytes := append(nonce, []byte(encryptedData)...)
+	checkSum := SumSHA256(nonceAndDataBytes)
+	var buf bytes.Buffer
+	_, err := buf.Write(checkSum[:])
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = buf.Write(nonceAndDataBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return buf.Bytes(), nil
+}
+
+// Decrypt decrypts the data
+func Decrypt(data []byte, password []byte) ([]byte, error) {
+	buf := bytes.NewBuffer(data)
+
+	checkSumBytes := make([]byte, blockSize)
+	n, err := buf.Read(checkSumBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	if n != blockSize {
+		return nil, errors.New("decode checksum failed")
+	}
+
+	var checkSum SHA256
+	copy(checkSum[:], checkSumBytes)
+
+	// verify the checksum
+	csh := SumSHA256(buf.Bytes())
+	if csh.Hex() != checkSum.Hex() {
+		return nil, errors.New("invalid checksum")
+	}
+
+	nonce := make([]byte, blockSize)
+	n, err = buf.Read(nonce)
+	if err != nil {
+		return nil, err
+	}
+
+	if n != blockSize {
+		return nil, errors.New("decode nonce failed")
+	}
+
+	encryptedBlocks := strings.Split(buf.String(), ",")
+	var decodeData []byte
+	for i := range encryptedBlocks {
+		b, err := base64.StdEncoding.DecodeString(encryptedBlocks[i])
+		if err != nil {
+			return nil, err
+		}
+
+		if len(b) != blockSize {
+			return nil, errors.New("invalid encrypted block string")
+		}
+
+		var bh SHA256
+		copy(bh[:], b[:])
+
+		dataHash := bh.Xor(hashPwdIndexNonce(password, int64(i), nonce))
+		decodeData = append(decodeData, dataHash[:]...)
+	}
+
+	buf = bytes.NewBuffer(decodeData)
+	l, err := binary.ReadUvarint(bytes.NewReader(decodeData[:lenPrefixSize]))
+	if err != nil {
+		return nil, ErrInvalidPassword
+	}
+
+	if l > uint64(len(decodeData[lenPrefixSize:])) || l < 32 {
+		return nil, ErrInvalidPassword
```

Comment:
is this `ErrInvalidPassword` or a different error?

Reply:
If the decrypted prefix length is not matched, we can say the password is invalid,  return "invalid prefix length" doesn't make sense to the user I think.

Reply:
Messages for the user should be transformed at the API layer, if necessary.  We do not care about the user in this method.